# Overview

## Objective

## Guiding Questions:
1. How can I ensure reproducibility without using the sort-of deprecated numpy random library?
2. Will setting a global seed through TensorFlow ensure reproducibility across runs if the operation level seeds are constant?
3. Which one is the better option in this project?

## Key findings
1. If possible, use the TensorFlow random generator.
2. Setting a global seed, and static operational seeds, ensures reproducibility, so far as I can see at least.
3. Irrelevant for now. I will go with the TF option unless I encounter a reason not to.

## Imports and setup

In [1]:
# must go first
%matplotlib inline
%config InlineBackend.figure_format='retina'

# Reloads functions each time so you can edit a script
# and not need to restart the kernel
%load_ext autoreload
%autoreload 2

# plotting
import matplotlib as mpl
from matplotlib import pyplot as plt
import seaborn as sns
sns.set_context("poster", font_scale=1.3)

import sys
import os
import datetime

sns.set()
sns.set_context('poster', font_scale=1.3)
sns.set_style("white")

import warnings
warnings.filterwarnings('ignore')

# basic wrangling
import numpy as np
import re
import pandas as pd

# tensorflow stuff
import tensorflow as tf

# Update matplotlib defaults to something nicer
mpl_update = {
    'font.size': 16,
    'xtick.labelsize': 14,
    'ytick.labelsize': 14,
    'figure.figsize': [12.0, 8.0],
    'axes.labelsize': 20,
    'axes.labelcolor': '#677385',
    'axes.titlesize': 20,
    'lines.color': '#0055A7',
    'lines.linewidth': 3,
    'text.color': '#677385',
    'font.family': 'sans-serif',
    'font.sans-serif': 'Tahoma'
}
mpl.rcParams.update(mpl_update)


Setting the global seed using `tf.random.set_seed(9001)`

In [2]:
tf.random.set_seed(9001)

In [3]:
rand_one = tf.random.uniform(shape=(2, 4), seed=42)
print(rand_one)

tf.Tensor(
[[0.49998724 0.5601554  0.4960822  0.04304016]
 [0.47926795 0.00552392 0.1810143  0.06998742]], shape=(2, 4), dtype=float32)


rand_one, round one
`tf.Tensor(`
`[[0.49998724 0.5601554  0.4960822  0.04304016]`
 `[0.47926795 0.00552392 0.1810143  0.06998742]], shape=(2, 4), dtype=float32)`

rand_one, round two
`tf.Tensor(`
`[[0.49998724 0.5601554  0.4960822  0.04304016]`
 `[0.47926795 0.00552392 0.1810143  0.06998742]], shape=(2, 4), dtype=float32)`


In [4]:
rand_two = tf.random.uniform(shape=(2, 4), seed=42)
print(rand_two)

tf.Tensor(
[[0.60508657 0.91349447 0.06239617 0.2568611 ]
 [0.6968026  0.16277087 0.13398087 0.69243324]], shape=(2, 4), dtype=float32)


rand_two, round one
`tf.Tensor(`
`[[0.60508657 0.91349447 0.06239617 0.2568611 ]`
 `[0.6968026  0.16277087 0.13398087 0.69243324]], shape=(2, 4), dtype=float32)`

rand_two, round two
`tf.Tensor(`
`[[0.60508657 0.91349447 0.06239617 0.2568611 ]`
 `[0.6968026  0.16277087 0.13398087 0.69243324]], shape=(2, 4), dtype=float32)`
